### ISSUE TITLE
[bug] Following Accounts Not Organized by Follow Date

### ISSUE THREAD

**Issue by CDN18** [superseriousbusiness/gotosocial#2769]
This issue arose from yesterday's discussion in the matrix room.

### Describe the bug with a clear and concise description of what the bug is.

Accounts in my Gotosocial Following list aren't organized by the order in when I followed them. This has been somewhat problematic for me, especially when trying to ascertain whether someone has accepted my follow request or, once accepted, difficulty in locating them in my Following list to add to a specific list.

It seems like an issue about `/api/v1/accounts/{id}/following`

### What's your GoToSocial Version?

v0.14.1-SNAPSHOT+git-51e8879

### GoToSocial Arch

arm64 Docker

### What happened?

Following accounts are not organized by follow date

### What you expected to happen?

Following accounts are sorted by follow date.

### How to reproduce it?

Simply visit oneself's GotoSocial profile and check the followings.

**Comment by daenney** [comment]
When loading them into the cache, we explicitly do so sorted by ID in descending order too. This should result in the newest ones always coming first: https://github.com/superseriousbusiness/gotosocial/blob/65b53660313f0faaa5d0bbb53ddcdb8ff23559d2/internal/db/bundb/relationship.go#L329-L335

Since our IDs are ULIDs and as such are temporally sortable this shouldn't be caused by clients doing some sorting either.

However, running the query myself `select * from follows where account_id = 'REDACTED' order by id DESC;`

```
id|created_at|updated_at|uri|account_id|target_account_id|show_reblogs|notify
```
```csv
FOLLOW_ID|2023-03-11 00:10:09|2023-03-11 00:10:09|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-09-20 18:07:34|2023-09-20 18:07:34|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-08-18 22:16:43|2023-10-14 23:39:41.903336+00:00|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|0|0
FOLLOW_ID|2023-05-04 12:39:09|2023-05-04 12:39:09|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2022-11-20 22:19:59|2023-07-16 05:48:50.792319+00:00|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|0|0
FOLLOW_ID|2023-04-30 22:28:39|2023-04-30 22:28:39|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-06-22 17:13:10|2023-06-22 17:13:10|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-06-29 19:47:11|2023-07-07 16:58:45.757367+00:00|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|0|0
FOLLOW_ID|2023-03-10 22:53:08|2023-03-10 22:53:08|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2022-12-29 12:27:46|2022-12-29 12:27:46|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-03-09 14:39:05|2023-03-09 14:39:05|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2022-11-21 13:19:04|2022-11-21 13:19:04|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-12-05 13:13:03|2023-12-05 13:13:03|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2022-12-13 21:52:40|2022-12-13 21:52:40|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2024-02-11 20:36:01|2024-02-11 23:20:05.513651+00:00|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|0|0
FOLLOW_ID|2023-06-23 22:27:45|2023-06-23 22:27:45|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-08-22 15:46:12|2023-08-22 15:46:12|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-10-23 19:50:40|2023-10-23 19:50:40|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-09-20 16:56:05|2023-09-20 16:56:05|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-08-29 22:43:46|2023-08-29 22:43:46|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-02-06 12:09:07|2023-02-06 12:09:07|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-06-28 10:49:36|2023-06-28 10:49:36|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-08-13 22:32:02|2023-08-13 22:32:02|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-07-17 19:31:59|2023-07-17 19:31:59|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2022-11-20 22:20:38|2022-11-20 22:20:38|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-10-09 16:48:17|2023-10-09 16:48:17|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-12-07 19:51:11|2023-12-07 19:51:11|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-02-27 13:17:06|2023-02-27 13:17:06|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2022-11-20 22:16:51|2022-11-20 22:16:51|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-07-27 18:34:32|2023-07-27 18:34:32|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-08-17 22:36:42|2023-08-17 22:36:42|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-11-15 13:20:56|2023-11-15 13:20:56|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2024-02-13 16:07:53|2024-02-13 16:07:53|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2022-11-20 22:12:43|2022-11-20 22:12:43|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-01-09 22:25:27|2023-01-09 22:25:27|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2022-11-26 08:56:52|2022-11-26 08:56:52|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-02-01 18:20:57|2023-02-01 18:20:57|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2024-01-10 08:51:25|2024-01-10 08:51:25|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-07-10 19:05:32|2023-07-10 19:05:32|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2024-01-08 10:36:28|2024-01-08 10:36:28|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2022-12-02 11:22:35|2022-12-02 11:22:35|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-10-09 22:07:34|2023-10-09 22:07:34|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2022-11-20 22:21:54|2022-11-20 22:21:54|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2022-11-23 10:05:44|2022-11-23 10:05:44|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-10-04 20:45:47|2023-10-04 20:45:47|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2024-03-18 14:42:10|2024-03-18 14:42:10|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2022-11-20 22:19:02|2022-11-20 22:19:02|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-02-04 11:04:50|2023-02-04 11:04:50|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-05-30 20:34:35|2023-05-30 20:34:35|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-05-26 16:00:06|2023-05-26 16:00:06|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-10-22 18:11:33|2023-10-22 18:11:33|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2024-01-11 12:39:37|2024-01-11 12:39:37|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-05-01 22:21:35|2023-05-01 22:21:35|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2022-11-20 22:20:56|2022-11-20 22:20:56|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2023-02-15 19:36:10|2023-02-15 19:36:10|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2022-11-20 22:21:18|2022-11-20 22:21:18|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
FOLLOW_ID|2024-03-13 16:33:15|2024-03-13 16:33:15|URI|ACCOUNT_ID|TARGET_ACCOUNT_ID|1|0
```

As we can see from the created and updated timestamps, this ordering is not what we want at all.

I've only tested this on SQLite, I don't have a PG. This may be a difference between how it does the sorting on a CHAR.

Also, something's off with some timestamps having nanoseconds and a timezone offset, but most not.

**Comment by daenney** [comment]
Doing the query without the `order by id DESC` does exactly what you expect. The first account I followed is at the top, the last one at the bottom and the created/updated timestamps are in chronologically ordered.

**Comment by daenney** [comment]
From @tsmethurst on Matrix:

> We don't always use temporally sortable IDs 
> Sometimes we use random ones 

That... might be our problem. If that's the case for the follows table, we need to sort on the created timestamp instead.

**Comment by daenney** [comment]
Indeed. The issue turns out to be that the IDs on the follows table are truly random instead of a ULID. As such sorting by `id` results in nonsense. The linked PR sorts by `created_at` instead, which should resolve the issue.

**Comment by daenney** [comment]
This will get released with 0.15.
